Provide the full Boost library to extension developers

alessandrogario opened the pull request:
This PR adds a small change to the Chocolatey and Homebrew scripts in order to build and ship all the Boost modules (on development environments) so that extension developers can use them without attempting to link a secondary Boost copy alongside the provided one

muffins commented on tools/provision.ps1:
```diff
@@ -346,7 +346,7 @@ function Install-ThirdParty {
   $packages = @(
     "aws-sdk-cpp.1.2.7",
-    "boost-msvc14.1.66.0",
+    "boost-msvc14.1.66.1",
```
For changes like these, where we're just updating something on our end and it's not an actual library update, I use a `-r1` or the like for the chocolatey version so we don't have versioning conflicts with subsequent boost iterations.

muffins commented:
**note** kicking off tests for this will cause a failure on Windows until we get the package uploaded, which I'll work on here shortly :)

muffins commented:
ok to test

osqueryer commented:
:-1: The commit a87f00dc81db241495adf38658d44c2e8ac4171a (Job results: [6997](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6997/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit a87f00dc81db241495adf38658d44c2e8ac4171a (Job results: [6998](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6998/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit dcaf63e8273a47047d0504456cc85175057f25b6 (Job results: [7156](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7156/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit dcaf63e8273a47047d0504456cc85175057f25b6 (Job results: [7157](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7157/)) failed one or more tests (**Linux**).

muffins commented:
ok to test

alessandrogario commented:
The `make deps` command works fine when run under Ubuntu 16.05 (desktop), and automatically builds Boost from source when it notices that the binary is not available.

I also tried on an Ubuntu 14.04 (server) and homebrew fails with the following error:

```
/usr/local/osquery/Library/Homebrew/utils/github.rb:226:in `raise_api_error': curl failed!  (GitHub::Error)
	from /usr/local/osquery/Library/Homebrew/utils/github.rb:184:in `open'
	from /usr/local/osquery/Library/Homebrew/utils/github.rb:233:in `issues_matching'
	from /usr/local/osquery/Library/Homebrew/utils/github.rb:272:in `issues_for_formula'
	from /usr/local/osquery/Library/Homebrew/exceptions.rb:369:in `fetch_issues'
	from /usr/local/osquery/Library/Homebrew/exceptions.rb:365:in `issues'
	from /usr/local/osquery/Library/Homebrew/exceptions.rb:420:in `dump'
	from /usr/local/osquery/Library/Homebrew/brew.rb:133:in `rescue in <main>'
	from /usr/local/osquery/Library/Homebrew/brew.rb:26:in `<main>'
```

I'm not really sure why it's failing, but I'll try to debug it!

alessandrogario commented:
I have managed to get a closer look at this error, and It seems to be normal that the compilation fails on Ubuntu 14.04. Even the current formula does not work, so the binary package we have on S3 now has been probably built on 16.04.

The `boostrap.sh` error is caused by the system linker being too old. The `raise_api_error` message is unrelated, and it's also probably caused by poor Ubuntu 14.04 support from homebrew.

To debug it, I've edited the `provision.sh` file adding the `--interactive` parameter at the end of the boost installer.

```
==> Installing boost from osquery/osquery-local
==> Downloading https://osquery-packages.s3.amazonaws.com/bottles/boost-1.66.0_300.x86_64_linux.bottle.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
Error: Failed to download resource "boost"
Download failed: https://osquery-packages.s3.amazonaws.com/bottles/boost-1.66.0_300.x86_64_linux.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading https://downloads.sourceforge.net/project/boost/boost/1.66.0/boost_1_66_0.tar.bz2
Already downloaded: /usr/local/osquery/.cache/boost-1.66.0.tar.bz2
==> Entering interactive mode
Type `exit' to return and finalize the installation
Install to this prefix: /usr/local/osquery/Cellar/boost/1.66.0_300

alessandro@tob-ubuntu1404-oldie:/tmp/boost-20180717-849-tpur45/boost_1_66_0$ ls
boost  boost-build.jam  boostcpp.jam  boost.css  boost.png  bootstrap.bat  bootstrap.sh  doc  index.htm  index.html  INSTALL  Jamroot  libs  LICENSE_1_0.txt  more  rst.css  status  tools

alessandro@tob-ubuntu1404-oldie:/tmp/boost-20180717-849-tpur45/boost_1_66_0$ ./bootstrap.sh --prefix=/usr/local/osquery/Cellar/boost/1.66.0_300 --with-toolset=cc
Building Boost.Build engine with toolset cc... 
Failed to build Boost.Build build engine
Consult 'bootstrap.log' for more details

alessandro@tob-ubuntu1404-oldie:/tmp/boost-20180717-849-tpur45/boost_1_66_0$ cat bootstrap.log
###
### Using 'cc' toolset.
###
rm -rf bootstrap
mkdir bootstrap
/usr/local/osquery/bin/clang -o bootstrap/jam0 -Os -w -pipe -march=native command.c compile.c constants.c debug.c execcmd.c frames.c function.c glob.c hash.c hdrmacro.c headers.c jam.c jambase.c jamgram.c lists.c make.c make1.c object.c option.c output.c parse.c pathsys.c regexp.c rules.c scan.c search.c subst.c timestamp.c variable.c modules.c strings.c filesys.c builtins.c class.c cwd.c native.c md5.c w32_getreg.c modules/set.c modules/path.c modules/regex.c modules/property-set.c modules/sequence.c modules/order.c execunix.c fileunix.c pathunix.c
/usr/bin/ld: /usr/local/osquery/opt/gcc/lib64/gcc/x86_64-unknown-linux-gnu/5.4.0/crtbegin.o: unrecognized relocation (0x2a) in section `.text'
/usr/bin/ld: final link failed: Bad value
clang-6.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

alessandrogario commented:
Some additional things I discovered: the Boost 1.60-r1 package for Windows that has been uploaded on S3 is broken. When compiling all components, it's easy to get errors about paths that are too long (I don't know if it's caused by Chocolatey, but I know there's an open bug about this).

The latest commit I added (`powershell: Add error checking in Invoke-BatchFile`) should catch this error not only for this case but for all .ps1 scripts.

I ended up cloning the repository inside `C:\Projects\osquery` (rather than inside my home) to solve this issue.

guliashvili commented:
@alessandrogario 
Thanks for your contribution. Are there any plans for correcting failing tests or should we close the PR?

alessandrogario commented:
Hey @guliashvili !

There's a summary of my findings here: https://github.com/facebook/osquery/pull/4339#issuecomment-405609518

This script has never worked (and probably never will) on Ubuntu 14.04, so it will keep failing until the binary package is uploaded to S3.

You can't build that package (even previous versions fail) because the system linker is too old on that OS. In order to merge this PR we just have to build the package and upload it so that homebrew finds it and stops trying to build it. I believe this is what has always been done when changing that script

muffins commented:
Yeah sorry, @obelisk and I both haven’t gotten around to uploading the package build yet, but if anyone else would like to we can get on merging this :)

muffins commented:
@alessandrogario can you swap out the hash for the linux bottle with `0930fdd0693f199c312cb7024442ffb86daedef4064297fedacfeee2a54a11fc`, and rebase, and we should be all set.

alessandrogario commented:
I have updated to release number of the boost-msvc14 package for Chocolatey, because the package was only partially built due to path length limitation issues (see 5d63256)

alessandrogario commented:
I've rebased on master; the Windows test is failing because I changed the package version and choco can no longer find the binary.

The reason I changed the version is because the package uploaded to S3 is broken due to path issues. if the working directory is too long the script will silently fail and build a package that doesn't contain all files. To build it correctly, I use `C:\Projects\osquery` as my root.

More info here: https://github.com/facebook/osquery/pull/4339#issuecomment-411727285

fmanco commented:
Given the change to Buck I assume this is now not needed anymore. @alessandrogario feel free to rebase on experimental and re-open otherwise.
